pkg/apis/core/types.go

```diff
@@ -3629,9 +3673,17 @@ type PodSecurityContext struct {
 	// no additional groups are added to any container. Note that group memberships
 	// defined in the container image for the uid of the container process are still effective,
 	// even if they are not included in this list.
+	// If SupplementalGroupsPolicy feature was enabled, you can control how the groups are calculated.
 	// Note that this field cannot be set when spec.os.name is windows.
 	// +optional
 	SupplementalGroups []int64
+	// Defines how supplemental groups of the first container processes are calculated.
+	// Valid values are "Merge" and "Strict". If not specified, "Merge" is used.
```

Comment:
Previously we said we cannot have a literal default value because of bugs, but once https://github.com/kubernetes/kubernetes/pull/124220 lands, we can actually do this properly.

Reply:
The bug today is that a new field with a default value gets considered by kubelet as a change, so it restarts all pods (and all nodes do this roughly in sync).  #124220  fixes that, so a new default should not trigger a restart.

> this field should be an OS dependent field

OK, thast may still be true, but adding a default will avoid code like:

```
if windows {
  return doSomethingElse()
}
policy = merge // hardcoded default
if FSGroupChangePolicy == nil {
    policy = *FSGroupChangePolicy
}
return doSomethingWithPolicy(policy)
```

And when users `kubectl get` the pods they will see "Merge" instead of nothing

Reply:
> so a new default should not trigger a restart

Wow, Thanks. That is why we have such constraint in `default_test.go`, right?

https://github.com/kubernetes/kubernetes/blob/5f1e5cafd33a8863c5dfe9f36c9cba836f0e981b/pkg/apis/core/v1/defaults_test.go#L165-L174

I didn’t know that. I’ve learned something today.

<hr />

> but adding a default will avoid code like:
> And when users kubectl get the pods they will see "Merge" instead of nothing

Thanks for this, too. Probably, now I got your idea. Once #124220 got landed, we can introduce to add new field with default value. But, this time, we would like to introduce conditional(os specific) default value like this:

- when `spec.os.name == windows`, it should validate the field must be `nil` 
- when `spec.os.name == linux`, it should make the default value be `Merge`

Then, perhaps, we can use custom defaulter function instead of `+default=“Merge”` annotation to achieve conditional default value?? I imagine like below:

```golang
# in pkg/apis/core/v1/defaults.go
func SetDefaults_PodSpec(obj *v1.PodSpec) {
…

	if obj.SecurityContext == nil {
		obj.SecurityContext = &v1.PodSecurityContext{}
	}
	if utilfeature.DefaultFeatureGate.Enabled(features.SupplementalGroupsPolicy) {
		if obj.SecurityContext.SupplementalGroupsPolicy == nil && (obj.OS == nil || obj.OS.Name == v1.Linux) {
			merge := v1.SupplementalGroupsPolicyMerge
			obj.SecurityContext.SupplementalGroupsPolicy = &merge
		}
	}
…
```

Reply:
@alexzielenski Here's a fun case for declarative defaults  :)

@everpeace - something like that, yes.

Reply:
> once https://github.com/kubernetes/kubernetes/pull/124220 lands, we can actually do this properly

well... once it lands and makes it to the oldest kubelet that could be running against the server (so once the n-3 kubelet is 1.31, so 1.34)

Reply:
oh, yeah, I guess that's a good point.  I got a little ahead of myself.  Sorry @everpeace - what we CAN do is leave "TODO" comments to revisit this in 1.34 :)